kamu-adapter-http/src/smart_protocol/ws_tungstenite_client.rs

```diff
@@ -320,6 +322,110 @@ impl WsSmartTransferProtocolClient {
         Ok(dataset_objects_push_response)
     }
 
+    async fn export_group_of_object_files(
+        &self,
+        socket: &mut TungsteniteStream,
+        push_objects_response: DatasetPushObjectsTransferResponse,
+        src: Arc<dyn Dataset>,
+        transfer_options: ObjectTransferOptions,
+    ) -> Result<(), SyncError> {
+        let total_files_count = push_objects_response.object_transfer_strategies.len();
+        let uploaded_files_counter = Arc::new(AtomicI32::new(0));
+        let cloned_counters: Vec<_> = iter::repeat_with(|| uploaded_files_counter.clone())
+            .take(total_files_count)
+            .collect();
+
+        let export_task = tokio::spawn(async move {
+            let src_ref = src.as_ref();
+            use futures::stream::{StreamExt, TryStreamExt};
+            futures::stream::iter(iter::zip(
+                push_objects_response.object_transfer_strategies,
+                cloned_counters,
+            ))
+            .map(Ok)
+            .try_for_each_concurrent(
+                /* limit */ transfer_options.max_parallel_transfers,
+                |(s, counter)| async move {
+                    let export_result = dataset_export_object_file(src_ref, &s).await;
+                    counter.fetch_add(1, Ordering::Relaxed);
+                    export_result
+                },
+            )
+            .await
+        });
+
+        while !export_task.is_finished() {
+            tokio::time::sleep(std::time::Duration::from_secs(
+                transfer_options.min_upload_progress_delay_sec,
+            ))
+            .await;
```

Comment:
A better approach that lets the process terminate faster could be something like:
```rust
loop {
  tokio::select! {
    _ = time::sleep(timeout) => ...,
    export_task => ...
  }
}
```

There is also [tokio::time::interval](https://docs.rs/tokio/latest/tokio/time/fn.interval.html) that might have less overhead than creating `sleep` task every time.